### PR TITLE
Support concurrent threads locking the same transaction

### PR DESCRIPTION
**Pull request by mystenmark**
Fix crash when two threads concurrently lock the same transaction.

The new test case fails reliably if the fix is not present.

### CODE REVIEW

File: crates/sui-core/src/execution_cache/object_locks.rs [hunk 1]
```diff
@@ -38,27 +38,19 @@ impl ObjectLocks {
...
         epoch_store: &AuthorityPerEpochStore,
     ) -> SuiResult<Option<LockDetails>> {
+        let tables = epoch_store.tables()?;
         match self.locked_transactions.entry(*obj_ref) {
-            DashMapEntry::Vacant(vacant) => {
-                let tables = epoch_store.tables()?;
+            DashMapEntry::Vacant(_) => {
                 let lock = tables.get_locked_transaction(obj_ref)?;
-                if let Some(lock_details) = lock {
-                    vacant.insert(lock_details);
```
Comment (mystenmark) [inline crates/sui-core/src/execution_cache/object_locks.rs:46]: warming the cache here could actually create a memory leak, removing

File: crates/sui-core/src/execution_cache/object_locks.rs [hunk 2]
```diff
@@ -38,27 +38,19 @@ impl ObjectLocks {
...
-                }
                 Ok(lock)
             }
             DashMapEntry::Occupied(occupied) => {
                 if cfg!(debug_assertions) {
-                    if let Some(lock_details) = epoch_store
-                        .tables()
-                        .unwrap()
```
Comment (mystenmark) [inline crates/sui-core/src/execution_cache/object_locks.rs:54]: cleaning up a bogus unwrap

File: crates/sui-core/src/execution_cache/object_locks.rs [hunk 3]
```diff
@@ -156,14 +151,17 @@ impl ObjectLocks {
     fn clear_cached_locks(&self, locks: &[(ObjectRef, LockDetails)]) {
         for (obj_ref, lock) in locks {
             let entry = self.locked_transactions.entry(*obj_ref);
-            let occupied = match entry {
+            let mut occupied = match entry {
                 DashMapEntry::Vacant(_) => panic!("lock must exist"),
```
Comment (lxfind) [inline crates/sui-core/src/execution_cache/object_locks.rs:155]: Any reason we can't just not panic here as the fix?

File: crates/sui-core/src/execution_cache/object_locks.rs [hunk 4]
```diff
@@ -23,7 +23,7 @@ pub(super) struct ObjectLocks {
     // those objects. Therefore we do a db read for each object we are locking.
     //
     // TODO: find a strategy to allow us to avoid db reads for each object.
-    locked_transactions: DashMap<ObjectRef, LockDetails>,
+    locked_transactions: DashMap<ObjectRef, (usize, LockDetails)>,
```
Comment (lxfind) [inline crates/sui-core/src/execution_cache/object_locks.rs:26]: Please add doc comment on the ref count

File: crates/sui-core/src/execution_cache/object_locks.rs (hunk 1 above)
Comment (lxfind) [reply crates/sui-core/src/execution_cache/object_locks.rs:46]: Please add this comment in code.

**Review by lxfind — APPROVED** [review]

File: crates/sui-core/src/execution_cache/object_locks.rs (hunk 3 above)
Comment (mystenmark) [reply crates/sui-core/src/execution_cache/object_locks.rs:155]: I will make it a debug panic only

File: crates/sui-core/src/execution_cache/object_locks.rs (hunk 1 above)
Comment (mystenmark) [reply crates/sui-core/src/execution_cache/object_locks.rs:46]: I'm actually going to go further and remove any reference to in memory state here to make it both simpler and clearer.

File: crates/sui-core/src/execution_cache/object_locks.rs (hunk 4 above)
Comment (mystenmark) [reply crates/sui-core/src/execution_cache/object_locks.rs:26]: added a type alias to make it clear

**Comment by halfprice** [comment]
Just to confirm:

It seems that this is when a user submit the same transaction multiple times, say tx_attempt1, tx_attempt2, and both locks are removed after `acquire_transaction_locks`. This lock is not the on disk object lock.

**Review by halfprice — APPROVED** [review]

**Comment by mystenmark** [comment]
> Just to confirm:
> 
> It seems that this is when a user submit the same transaction multiple times, say tx_attempt1, tx_attempt2, and both locks are removed after `acquire_transaction_locks`. This lock is not the on disk object lock.

The in memory state is just a way to ensure (without relying on rocksdb transactions) that when two transactions contend for a lock, one of them succeeds and the other writes no locks at all. The on-disk state is the authoritative record of locked objects.

File: crates/sui-core/src/execution_cache/object_locks.rs (hunk 3 above)
Comment (lxfind) [reply crates/sui-core/src/execution_cache/object_locks.rs]: If we don't panic here, do we even need ref count?

**Pull request merged**